File: solutions/0647. Palindromic Substrings/0647.py
# Time complexity: O(n^2)
# Space complexity: O(1)
class Solution:
  def countSubstrings(self, s: str) -> int:
    ans = 0

    for i in range(len(s)):
      ans += self.extendPalindromes(s, i, i)
      ans += self.extendPalindromes(s, i, i + 1)

    return ans

  def extendPalindromes(self, s: str, left: int, right: int) -> int:
    count = 0

    while left >= 0 and right < len(s) and s[left] == s[rigth]:
      count += 1
      left -= 1
      right += 1

    return count

File: solutions/0647. Palindromic Substrings/0647.cpp
// Time complexity: O(n^2)
// Space complexity: O(1)
class Solution {
public:
  int countSubstrings(string_view s) {
    int ans = 0;

    for (int i = 0; i < s.length(); ++i) {
      ans += extendPalindromes(s, i, i);
      ans += extendPalindromes(s, i, i + 1);
    }

    return ans;
    }
private:
  int extendPalindromes(string_view s, int left, int right) {
    int count = 0;

    while (left >= 0 && right < s.length() && s[left] == s[right]) {
      ++count;
      --left;
      ++right;
    }

    return count;
  }    
};